File: src/main/java/com/medicalsystem/factory/SectionFactory.java
package com.medicalsystem.factory;

import com.medicalsystem.model.FormType;
import com.medicalsystem.model.Section;
import com.medicalsystem.model.field.Field;
import com.medicalsystem.properties.ConfigProperties;
import com.medicalsystem.service.FieldService;
import lombok.extern.java.Log;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
@Log
public class SectionFactory {

    private static FieldService fieldService;

    @Autowired
    public SectionFactory(FieldService fieldService) {
        SectionFactory.fieldService = fieldService;
    }

    /**
     * Creates a Section object from the config file
     *
     * @param _section an object representing a section loaded from config
     * @return         created Section object
     */
    public static Section fromConfig(ConfigProperties.Section _section, FormType formType) {
        Section section = new Section();

        // Set name
        section.setName(_section.getName());

        // Create and add fields
        _section.getFields().stream()
                .filter(_field -> fieldIsRelevant(_field, formType))
                .forEach(_field -> {
                    Field field = FieldFactory.fromConfig(_field);
                    fieldService.saveOrUpdate(field);
                    section.addField(field);
                });

        log.info(String.format("Section created: '%s'", section.getName()));

        return section;
    }

    private static boolean fieldIsRelevant(ConfigProperties.Section.Field field, FormType formType) {

        if (formType == FormType.OPEN)
            return field.getOpenIndex() != -1;

        if (formType == FormType.EVAR)
            return field.getEvarIndex() != -1;

        return false;
    }

}

File: src/main/kotlin/com/medicalsystem/factory/SectionFactory.kt
package com.medicalsystem.factory

import com.medicalsystem.model.Section
import com.medicalsystem.model.dto.SectionDTO
import com.medicalsystem.properties.FormProperties.PropForm.PropSection

object SectionFactory : PropertiesFactory<PropSection, Section>, DTOFactory<SectionDTO, Section> {

    override fun createFromProperties(t: List<PropSection>): List<Section> =
            t.map(this::createFromProperties)

    override fun createFromProperties(t: PropSection): Section {
        val section = Section(name = t.name)
        val fields = FieldFactory.createFromProperties(t.fields)

        section.fields = fields.toMutableList()
        fields.forEach { field -> field.section = section }

        return section
    }

    override fun createEmptyDTO(us: List<Section>): List<SectionDTO> =
            us.map { createEmptyDTO(it) }

    override fun createEmptyDTO(u: Section): SectionDTO =
            SectionDTO(
                    id = u.id,
                    name = u.name,
                    fields = FieldFactory.createEmptyDTO(u.fields)
            )

    override fun createEmptyFromDTO(ts: List<SectionDTO>): List<Section> =
            ts.map { createEmptyFromDTO(it) }

    override fun createEmptyFromDTO(t: SectionDTO): Section =
            Section(
                    id = t.id,
                    name = t.name
            )
}